File: app/src/main/java/com/mxj/mmitest/ui/testitems/SimTestActivity.java
package com.mxj.mmitest.ui.testitems;

import android.Manifest;
import android.content.Context;
import android.os.Bundle;
import android.telephony.TelephonyManager;
import android.widget.LinearLayout;
import android.widget.TextView;
import com.mxj.mmitest.data.repository.TestRepository;
import com.mxj.mmitest.ui.base.BaseTestActivity;

/**
 * SIM卡测试
 */
public class SimTestActivity extends BaseTestActivity {

    private TestRepository repository;
    private TelephonyManager telephonyManager;
    private LinearLayout contentLayout;
    private boolean hasSim = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        repository = TestRepository.getInstance(this);
        telephonyManager = (TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE);

        setupContentView();

        super.onCreate(savedInstanceState);
    }

    private void setupContentView() {
        contentLayout = new LinearLayout(this);
        contentLayout.setOrientation(LinearLayout.VERTICAL);
        contentLayout.setPadding(48, 32, 48, 32);

        updateSimDisplay();

        setContentView(contentLayout);
    }

    private void updateSimDisplay() {
        contentLayout.removeAllViews();

        TextView titleView = new TextView(this);
        titleView.setText("SIM卡检测");
        titleView.setTextSize(20);
        titleView.setTextColor(0xFF000000);
        contentLayout.addView(titleView);

        try {
            String simState = telephonyManager.getSimState() + "";
            int simStateInt = telephonyManager.getSimState();

            TextView stateView = new TextView(this);
            stateView.setText("\nSIM卡状态: " + getSimStateText(simStateInt));
            stateView.setTextSize(16);
            stateView.setTextColor(0xFF333333);
            contentLayout.addView(stateView);

            // 检查是否有SIM卡
            if (simStateInt == TelephonyManager.SIM_STATE_READY) {
                hasSim = true;

                TextView operatorView = new TextView(this);
                String operatorName = telephonyManager.getSimOperatorName();
                operatorView.setText("\n运营商: " + (operatorName != null ? operatorName : "未知"));
                operatorView.setTextSize(14);
                operatorView.setTextColor(0xFF333333);
                contentLayout.addView(operatorView);

                TextView infoView = new TextView(this);
                infoView.setText("\nSIM卡正常识别，可以进行通话和上网");
                infoView.setTextSize(14);
                infoView.setTextColor(0xFF4CAF50);
                contentLayout.addView(infoView);
            } else {
                hasSim = false;
                TextView noSimView = new TextView(this);
                noSimView.setText("\n未检测到SIM卡或SIM卡不可用\n请检查SIM卡是否正确插入");
                noSimView.setTextSize(16);
                noSimView.setTextColor(0xFFF44336);
                contentLayout.addView(noSimView);
            }
        } catch (SecurityException e) {
            hasSim = false;
            TextView errorView = new TextView(this);
            errorView.setText("\n需要READ_PHONE_STATE权限\n请授予权限后重试");
            errorView.setTextSize(14);
            errorView.setTextColor(0xFFF44336);
            contentLayout.addView(errorView);
        } catch (Exception e) {
            hasSim = false;
            TextView errorView = new TextView(this);
            errorView.setText("\n检测失败: " + e.getMessage());
            errorView.setTextSize(14);
            errorView.setTextColor(0xFFF44336);
            contentLayout.addView(errorView);
        }

        setPassEnabled(hasSim);
    }

    private String getSimStateText(int state) {
        switch (state) {
            case TelephonyManager.SIM_STATE_ABSENT:
                return "无SIM卡";
            case TelephonyManager.SIM_STATE_READY:
                return "就绪";
            case TelephonyManager.SIM_STATE_PIN_REQUIRED:
                return "需要PIN码";
            case TelephonyManager.SIM_STATE_PUK_REQUIRED:
                return "需要PUK码";
            case TelephonyManager.SIM_STATE_NETWORK_LOCKED:
                return "网络锁定";
            case TelephonyManager.SIM_STATE_UNKNOWN:
            default:
                return "未知状态";
        }
    }

    @Override
    protected String getTestName() {
        return "SIM卡测试";
    }

    @Override
    protected String getTestDescription() {
        return "请检查SIM卡是否正常识别\n\n操作步骤：\n1. 确认SIM卡已插入\n2. 检查信号强度\n3. 点击PASS或FAIL按钮";
    }

    @Override
    protected int getTimeoutSeconds() {
        return 15;
    }

    @Override
    protected String[] getRequiredPermissions() {
        return new String[]{Manifest.permission.READ_PHONE_STATE};
    }

    @Override
    protected void onTestExecute() {
        updateSimDisplay();
    }

    @Override
    protected boolean isPassEnabled() {
        return hasSim;
    }

    @Override
    protected void finishAndSaveResult(boolean passed) {
        repository.saveSingleTestResultSync(1, "SIM卡测试", passed, getDeviceUniqueId());
        super.finishAndSaveResult(passed);
    }
}

File: app/src/main/java/com/mxj/mmitest/ui/testitems/SimTestActivity.kt
package com.mxj.mmitest.ui.testitems

import android.os.Bundle
import android.telephony.SubscriptionManager
import android.telephony.TelephonyManager
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.core.content.getSystemService
import androidx.lifecycle.lifecycleScope
import com.mxj.mmitest.data.repository.TestRepository
import com.mxj.mmitest.ui.base.BaseActivity
import com.mxj.mmitest.ui.components.TestItemScreen
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * SIM卡测试Activity
 */
class SimTestActivity : BaseActivity() {

    private val testName = "SIM卡测试"
    private val testItemId = 1
    private val timeoutSeconds = 15

    private lateinit var repository: TestRepository
    private lateinit var telephonyManager: TelephonyManager
    private lateinit var subscriptionManager: SubscriptionManager

    // SIM卡信息数据类
    data class SimCardInfo(
        val slotIndex: Int,
        val displayName: String,
        val carrierName: String,
        val simSerialNumber: String,
        val isReady: Boolean,
        val simState: String
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        repository = TestRepository(this)
        telephonyManager = getSystemService<TelephonyManager>()!!
        subscriptionManager = getSystemService<SubscriptionManager>()!!

        setContent {
            var remainingSeconds by remember { mutableIntStateOf(timeoutSeconds) }
            var passEnabled by remember { mutableStateOf(false) }
            var simCards by remember { mutableStateOf<List<SimCardInfo>>(emptyList()) }

            // 获取SIM卡信息
            LaunchedEffect(Unit) {
                simCards = getSimCardInfoList()
                // 只要有任意一张SIM卡Ready，就允许PASS
                passEnabled = simCards.any { it.isReady }
            }

            TestItemScreen(
                testName = testName,
                testDescription = "请检查SIM卡是否正常识别\n\n" +
                        "操作步骤：\n" +
                        "1. 确认SIM卡已插入\n" +
                        "2. 检查信号强度\n" +
                        "3. 点击PASS或FAIL按钮",
                remainingSeconds = remainingSeconds,
                onPass = { saveAndFinish(true) },
                onFail = { saveAndFinish(false) },
                passEnabled = passEnabled,
                content = {
                    // 自定义内容：SIM卡信息列表
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(16.dp)
                    ) {
                        if (simCards.isEmpty()) {
                            Text(
                                text = "未检测到SIM卡",
                                style = MaterialTheme.typography.bodyLarge,
                                color = Color.Gray
                            )
                        } else {
                            simCards.forEach { simInfo ->
                                SimCardInfoRow(simInfo = simInfo)
                                Spacer(modifier = Modifier.height(16.dp))
                            }
                        }
                    }
                }
            )

            // 倒计时逻辑 - 超时直接finish，不弹框
            LaunchedEffect(Unit) {
                while (remainingSeconds > 0) {
                    delay(1000)
                    remainingSeconds--
                }
                // 超时自动结束，标记为FAIL
                saveAndFinish(false)
            }
        }
    }

    // 获取所有SIM卡信息列表
    private fun getSimCardInfoList(): List<SimCardInfo> {
        val simList = mutableListOf<SimCardInfo>()

        try {
            val activeSubscriptionInfoList = subscriptionManager.activeSubscriptionInfoList
            if (activeSubscriptionInfoList != null) {
                for (subscriptionInfo in activeSubscriptionInfoList) {
                    val slotIndex = subscriptionInfo.simSlotIndex
                    val displayName = subscriptionInfo.displayName?.toString() ?: "SIM $slotIndex"
                    val carrierName = subscriptionInfo.carrierName?.toString() ?: "未知运营商"
                    val simSerialNumber = subscriptionInfo.iccId ?: "未知"

                    // 获取该卡槽的SimState来判断是否就绪
                    val simStateRaw = telephonyManager.getSimState(slotIndex)
                    val isReady = simStateRaw == 5  // SIM_STATE_READY
                    val simState = if (isReady) "正常" else "未就绪"

                    simList.add(
                        SimCardInfo(
                            slotIndex = slotIndex,
                            displayName = displayName,
                            carrierName = carrierName,
                            simSerialNumber = simSerialNumber,
                            isReady = isReady,
                            simState = simState
                        )
                    )
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        // 如果没有获取到任何SIM卡信息，尝试使用传统方式检测
        if (simList.isEmpty()) {
            for (slot in 0 until telephonyManager.phoneCount) {
                val simStateRaw = telephonyManager.getSimState(slot)
                val stateStr = when (simStateRaw) {
                    5 -> "正常"  // SIM_STATE_READY
                    1 -> "不存在"  // SIM_STATE_ABSENT
                    2, 3 -> "锁定"  // SIM_STATE_LOCKED_PIN or SIM_STATE_LOCKED_PUK
                    else -> "未知"
                }
                val isReady = simStateRaw == 5  // SIM_STATE_READY
                simList.add(
                    SimCardInfo(
                        slotIndex = slot,
                        displayName = "SIM ${slot + 1}",
                        carrierName = if (isReady) "已识别" else "-",
                        simSerialNumber = "-",
                        isReady = isReady,
                        simState = stateStr
                    )
                )
            }
        }

        return simList
    }

    @Composable
    private fun SimCardInfoRow(simInfo: SimCardInfo) {
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = if (simInfo.isReady) Color(0xFF1B5E20).copy(alpha = 0.1f) else Color(0xFFB71C1C).copy(alpha = 0.1f)
            )
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp)
            ) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = simInfo.displayName,
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color = if (simInfo.isReady) Color(0xFF1B5E20) else Color(0xFFB71C1C)
                    )
                    Text(
                        text = simInfo.simState,
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.Bold,
                        color = if (simInfo.isReady) Color(0xFF4CAF50) else Color(0xFFF44336)
                    )
                }

                Spacer(modifier = Modifier.height(8.dp))

                InfoRow(label = "运营商", value = simInfo.carrierName)
                Spacer(modifier = Modifier.height(4.dp))
                InfoRow(label = "卡槽", value = "${simInfo.slotIndex + 1}")
                Spacer(modifier = Modifier.height(4.dp))
                InfoRow(label = "ICCID", value = simInfo.simSerialNumber)
            }
        }
    }

    @Composable
    private fun InfoRow(label: String, value: String) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = label,
                style = MaterialTheme.typography.bodyMedium,
                color = Color.Gray
            )
            Text(
                text = value,
                style = MaterialTheme.typography.bodyMedium
            )
        }
    }

    private fun saveAndFinish(passed: Boolean) {
        lifecycleScope.launch {
            repository.saveSingleTestResult(
                testItemId = testItemId,
                testItemName = testName,
                passed = passed,
                deviceId = android.provider.Settings.Secure.getString(
                    contentResolver,
                    android.provider.Settings.Secure.ANDROID_ID
                ) ?: android.os.Build.MODEL
            )
            finish()
        }
    }
}
